File: .claude/hooks/tenant_guard.py
import json
import re
import sys

data = json.load(sys.stdin)
tool_input = data.get("tool_input", {})
file_path = tool_input.get("file_path", "")

if not file_path.endswith(".java") or file_path.endswith("TenantConnectionListener.java"):
    sys.exit(0)

content = tool_input.get("content") or tool_input.get("new_string") or ""
if not content:
    sys.exit(0)

raw_jdbc = re.search(r"\b(DriverManager\.getConnection|java\.sql\.Connection\b|\.createStatement\(|\.prepareStatement\()", content)
if raw_jdbc and "DSLContext" not in content:
    print(json.dumps({
        "systemMessage": (
            "Raw JDBC pattern detected outside TenantConnectionListener in "
            f"{file_path}. CLAUDE.md rule: every transaction must SET LOCAL app.clinic_id "
            "before any query, or RLS silently returns empty results. Use DSLContext / "
            "@Transactional instead, or confirm this path binds TenantContext first."
        )
    }))

sys.exit(0)

File: .opencode/plugins/tenant-guard.ts
import type { Plugin } from "@opencode-ai/plugin"

const RAW_JDBC = /\b(DriverManager\.getConnection|java\.sql\.Connection\b|\.createStatement\(|\.prepareStatement\()/

const tenantGuard: Plugin = async () => {
  return {
    "tool.execute.before": async (input, output) => {
      const tool = String(input?.tool ?? "").toLowerCase()
      if (tool !== "edit" && tool !== "write") return

      const args = output?.args ?? {}
      const filePath = String(args.filePath ?? "").replaceAll("\\", "/")
      if (filePath.endsWith("TenantConnectionListener.java")) return
      if (!filePath.endsWith(".java")) return

      const content = String(args.content ?? args.newString ?? "")
      if (!content) return
      if (!RAW_JDBC.test(content)) return
      if (content.includes("DSLContext")) return

      throw new Error(
        `CLINICOS TENANT GUARD: raw JDBC pattern detected in ${args.filePath}. ` +
          "Every transaction must set app.clinic_id via TenantContext before any query, " +
          "or RLS silently returns empty results. Use DSLContext / @Transactional instead, " +
          "or confirm this path binds TenantContext first.",
      )
    },
  }
}

export default tenantGuard